### PR TITLE
fix: [CDS-87157]: Refactors logic to evaluate build detail options to ensure stale data is not used by component

### PR DESCRIPTION
**Pull request by samarth-sharma31**
### Summary

This PR refactors logic to evaluate build detail options as a memoized value and not as a function. This will ensure that stale closure does not get created for the child component using the build detail options.

#### Screenshots

https://github.com/harness/harness-core-ui/assets/96409598/d1b45eb4-ad28-454a-9025-54f8f82be644

#### PR Checklist

<details>
<summary>Please check if your PR fulfils the following requirements</summary>

- [ ] Tests for the changes have been added. Ideally, include a test that fails without this PR but passes with it.
- [ ] Docs have been [added/updated](https://harness.atlassian.net/jira/software/c/projects/DOC/boards/40).
</details>

<details>
<summary>Use the following comments to re-trigger PR Checks</summary>

- Jest: `retrigger jest`
- Prettier: `retrigger prettier`
- Type Check: `retrigger typecheck`
- ESLint: `retrigger eslint`
- Sonar: `retrigger sonar`
- Standards: `retrigger standards`
- Build: `retrigger build`
- Title Check: `retrigger titlecheck`
- Feature Name Check: `trigger featurenamecheck`
- Coverage: `retrigger coverage`
- Rebase: `trigger rebase`
- Cypress Rest: `retrigger cypress-rest`
- Cypress CD: `retrigger cypress-cd`
- Cypress Pipeline: `retrigger cypress-pipeline`
- Cypress CV: `retrigger cypress-cv`
- Fix Prettier: `fix prettier`
- MessageMetadata: `trigger messagecheck`
- GitLeaks: `trigger gitLeaks`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

### Rebase Instructions:
To rebase your branch onto the latest changes in the target branch, simply use the following command in a comment: `/rebase`

### CODE REVIEW

**Comment by samarth-sharma31** [comment]
trigger messagecheck

File: src/modules/75-cd/components/PipelineSteps/K8sServiceSpec/ArtifactSource/CustomArtifactSource/CustomArtifactSource.tsx [hunk 1]
```diff
@@ -191,19 +191,17 @@ const Content = (props: CustomArtifactRenderContent): React.ReactElement => {
...
-  }, [buildDetails?.data])
 
-  const getBuildDetails = (): { label: string; value: string }[] => {
-    if (fetchingBuilds) {
-      return [{ label: 'Loading Builds...', value: 'Loading Builds...' }]
-    }
-    return defaultTo(selectItems, [])
-  }
+    return defaultTo(buildItems, [])
+  }, [buildDetails?.data, fetchingBuilds]) as { label: string; value: string }[]
```
Comment (PratyushGar) [inline src/modules/75-cd/components/PipelineSteps/K8sServiceSpec/ArtifactSource/CustomArtifactSource/CustomArtifactSource.tsx:204]:
```suggestion
  }, [buildDetails?.data, fetchingBuilds]) as SelectOption[]
```

File: src/modules/75-cd/components/PipelineSteps/K8sServiceSpec/ArtifactSource/CustomArtifactSource/CustomArtifactSource.tsx [hunk 2]
```diff
@@ -191,19 +191,17 @@ const Content = (props: CustomArtifactRenderContent): React.ReactElement => {
     )
   )
 
-  const selectItems = useMemo(() => {
-    return buildDetails?.data?.map((builds: BuildDetails) => ({
+  const buildDetailOptions = useMemo(() => {
+    if (fetchingBuilds) {
+      return [{ label: 'Loading Builds...', value: 'Loading Builds...' }]
```
Comment (PratyushGar) [inline src/modules/75-cd/components/PipelineSteps/K8sServiceSpec/ArtifactSource/CustomArtifactSource/CustomArtifactSource.tsx:196]: nit: can we move this to strings

File: src/modules/75-cd/components/PipelineSteps/K8sServiceSpec/ArtifactSource/CustomArtifactSource/CustomArtifactSource.tsx (hunk 1 above)
Comment (samarth-sharma31) [reply src/modules/75-cd/components/PipelineSteps/K8sServiceSpec/ArtifactSource/CustomArtifactSource/CustomArtifactSource.tsx]: done

File: src/modules/75-cd/components/PipelineSteps/K8sServiceSpec/ArtifactSource/CustomArtifactSource/CustomArtifactSource.tsx (hunk 2 above)
Comment (samarth-sharma31) [reply src/modules/75-cd/components/PipelineSteps/K8sServiceSpec/ArtifactSource/CustomArtifactSource/CustomArtifactSource.tsx]: agreed, done

**Review by PratyushGar — APPROVED** [review]

**Review by namanharness — APPROVED** [review]

**Pull request merged**